### PR TITLE
feat: Filter by category

### LINKED ISSUE #726: request: Filter by category
### Is there an existing feature request for this?

- [X] I have searched the existing issues.

### Command

_No response_

### Description

As a developer, I want to run specific commands in specific projects based on the category that they are added to.

### Reasoning

This feature can be useful in contexts where you have specific scripts to run only in a few packages. For Instance, looking at the following structure, if I'd like to run a script in a and b, I need to use `melos exec --scope="a" --scope="b"` or other filters, but actually, I cannot run it based on a category, like packages part of lint feature.

- packages
  - a
  - b
  - c

### Additional context and comments

In my opinion, it could be added to the `melos.yml` file, such as:
```yml
# ...

categories:
  category_a:
    - packages/a
    - packages/c

# ...
```

Because of this, the filter `--category` can be created.

### PR DESCRIPTION
**Pull request by tilucasoli**
## Description
closes #726
- I added a new type of filter, now it's possible to filter packages by category.

## Type of Change

- [x] ✨ `feat` -- New feature (non-breaking change which adds functionality)
- [ ] 🛠️ `fix` -- Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ `!` -- Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 `refactor` -- Code refactor
- [ ] ✅ `ci` -- Build configuration change
- [ ] 📝 `docs` -- Documentation
- [ ] 🗑️ `chore` -- Chore

### CODE REVIEW

File: packages/melos/lib/src/package.dart [hunk 1]
```diff
@@ -379,6 +396,7 @@ class PackageFilters {
     String? diff,
     bool? includeDependencies,
     bool? includeDependents,
+    List<Glob>? category,
```
Comment (spydon) [inline packages/melos/lib/src/package.dart:399]: Follow the same convention as the rest of the fields here and use the same name as in the constructor

File: packages/melos/lib/src/package.dart [hunk 2]
```diff
@@ -747,6 +786,18 @@ extension on Iterable<Package> {
     }).toList();
   }
 
+  Iterable<Package> applyCategory(List<Glob> appliedCategories) {
```
Comment (spydon) [inline packages/melos/lib/src/package.dart:789]:
```suggestion
  Iterable<Package> applyCategories(List<Glob> appliedCategories) {
```

File: packages/melos/lib/src/package.dart [hunk 3]
```diff
@@ -1134,3 +1187,10 @@ class Plugin {
   Map<Object?, Object?>? get platforms =>
       _plugin['platforms'] as Map<Object?, Object?>?;
 }
+
+class PackageCategory {
```
Comment (spydon) [inline packages/melos/lib/src/package.dart:1191]: Is this used anywhere?

File: packages/melos/lib/src/workspace_configs.dart [hunk 4]
```diff
@@ -317,6 +342,14 @@ class MelosWorkspaceConfig {
       name: name,
       repository: repository,
       sdkPath: sdkPath,
+      categories: categories.map(
+        (key, value) => MapEntry(
+          key,
+          value.map((v) {
+            return Glob(v);
+          }).toList(),
```
Comment (spydon) [inline packages/melos/lib/src/workspace_configs.dart:350]:
```suggestion
          value.map(glob.new).toList(),
```

**Review by spydon — COMMENTED** [review]
Some minor comments, but overall it looks good!
It's missing docs though, the docs live here:
https://github.com/invertase/melos/blob/main/docs

File: packages/melos/test/package_test.dart [hunk 5]
```diff
@@ -239,6 +240,57 @@ void main() {
...
+            path: '/test',
+            pathRelativeToWorkspace: 'test',
+            version: Version(1, 0, 0),
+            publishTo: Uri(),
+            pubSpec: const PubSpec(),
+            categories: category,
+          );
+        }
+
+        final sut = [
```
Comment (spydon) [inline packages/melos/test/package_test.dart:262]: Last nit, use a better name for this variable

**Review by spydon — APPROVED** [review]
Thanks for your contribution!

File: docs/configuration/overview.mdx [hunk 6]
```diff
@@ -110,10 +110,24 @@ ignore:
...
 
-<Info>
-  You can also expand the scope of packages on a per-command basis via the
-  [`--scope` filter](/filters#--scope) flag.
-</Info>
+## categories
+
+Categories are used to group packages together.
+
+To define a custom package categories, add a `categories` section in your `melos.yaml` file.
```
Comment (spydon) [inline docs/configuration/overview.mdx:117]:
```suggestion
To define custom package categories, add a `categories` section in your `melos.yaml` file.
```

**Comment by tilucasoli** [comment]
I pushed a new commit with an update for the documentation. would you take a look to confirm if it is ok?

File: docs/filters.mdx [hunk 7]
```diff
@@ -46,6 +46,15 @@ repeated.
 melos exec --ignore="*internal*" -- flutter build ios
 ```
 
+## --categories
+
+Filter packages based on categories declared in the `melos.yaml` file.
+
+```bash
+# Run `flutter build ios` on all packages but ignore those whose packages names contain "internal"
```
Comment (spydon) [inline docs/filters.mdx:54]: This one is not updated

**Pull request merged**